core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java

```diff
@@ -920,25 +968,34 @@ public void onFailure(Throwable t)
     /**
      * Move the task directly to the failed state if there was a failure in this task
      */
-    @Override
-    public synchronized void fail(Throwable cause)
+    private synchronized void fatalUnacknowledgedFailure(Throwable cause)
     {
-        TaskStatus taskStatus = getTaskStatus();
-        if (!taskStatus.getState().isDone()) {
-            log.debug(cause, "Remote task %s failed with %s", taskStatus.getSelf(), cause);
-        }
-
-        TaskStatus status = failWith(getTaskStatus(), FAILED, ImmutableList.of(toFailure(cause)));
-        taskStatusFetcher.updateTaskStatus(status);
-
         try (SetThreadName ignored = new SetThreadName("HttpRemoteTask-%s", taskId)) {
-            if (cause instanceof TrinoTransportException) {
-                // task is unreachable
-                cleanUpLocally();
-            }
-            else {
-                // send abort to task
-                scheduleAsyncCleanupRequest(new Backoff(maxErrorDuration), "abort", true);
+            TaskStatus taskStatus = getTaskStatus();
+            if (!taskStatus.getState().isDone()) {
+                // Update the taskInfo with the new taskStatus.
+
+                // Generally, we send a cleanup request to the worker, and update the TaskInfo on
+                // the coordinator based on what we fetched from the worker. If we somehow cannot
+                // get the cleanup request to the worker, the TaskInfo that we fetch for the worker
+                // likely will not say the task is done however many times we try. In this case,
+                // we have to set the local query info directly so that we stop trying to fetch
+                // updated TaskInfo from the worker. This way, the task on the worker eventually
+                // expires due to lack of activity.
+                List<ExecutionFailureInfo> failures = ImmutableList.<ExecutionFailureInfo>builderWithExpectedSize(taskStatus.getFailures().size() + 1)
+                        .add(toFailure(cause))
+                        .addAll(taskStatus.getFailures())
+                        .build();
+                taskStatus = failWith(taskStatus, FAILED, failures);
+                if (cause instanceof TrinoTransportException) {
+                    // Since this TaskInfo is updated in the client the "complete" flag will not be set,
```

Comment:
What is the "complete" flag?

Reply:
Haha, that's a great question. During the (fairly significant) refactoring here I was trying to preserve the original comments where possible, but you're right that actually there is indeed no "complete" flag anymore (was there ever?) so the comment is wrong. That does beg the question, _should_ there be a complete flag? I could add one to differentiate a "locally failed immediately" from "acknowledged termination" but no code would be using that at the moment.